flow/BlobCipher.cpp

```diff
@@ -0,0 +1,392 @@
+                                              const int plaintextLen,
+                                              BlobCipherEncryptHeader* header,
+                                              Arena& arena) {
+	TEST(true); // Encrypting data with BlobCipher
+	auto ciphertext = new (arena) unsigned char[plaintextLen + AES_BLOCK_SIZE];
+	int bytes{ 0 };
+	if (EVP_EncryptUpdate(ctx, ciphertext, &bytes, plaintext, plaintextLen) != 1) {
+		throw internal_error();
+	}
+	int finalBytes{ 0 };
+	if (EVP_EncryptFinal_ex(ctx, ciphertext + bytes, &finalBytes) != 1) {
+		throw internal_error();
+	}
+	// populate header details for the encrypted blob.
+	header->flags.size = sizeof(BlobCipherEncryptHeader);
+	header->flags.headerVersion = EncryptBlobCipherAes265Ctr::ENCRYPT_HEADER_VERSION;
+	header->flags.encryptMode = BLOB_CIPHER_ENCRYPT_MODE_AES_256_CTR;
+	header->baseCipherId = cipherKey->getBaseCipherId();
+	header->encryptDomainId = cipherKey->getDomainId();
+	header->salt = cipherKey->getSalt();
+
+	// Preserve checksum of encrypted bytes in the header; approach protects against disk induced bit-rot/flip
+	// scenarios. AES CTR mode doesn't generate 'tag' by default as with schemes such as: AES 256 GCM.
+	//
+	// Possible to use XXH3_64bitsWithSeed() where 'seed' could be cipherKey->randomSalt; for performance reasons
+	// choosing 'default seed'.
+
+	header->checksum = XXH3_64bits(ciphertext, bytes + finalBytes);
+	return StringRef(ciphertext, bytes + finalBytes);
+}
+
+EncryptBlobCipherAes265Ctr::~EncryptBlobCipherAes265Ctr() {
+	EVP_CIPHER_CTX_free(ctx);
+}
+
+// DecryptBlobCipher class methods
+
+DecryptBlobCipherAes256Ctr::DecryptBlobCipherAes256Ctr(Reference<BlobCipherKey> key, const BlobCipherIV& iv)
+  : ctx(EVP_CIPHER_CTX_new()) {
+	if (ctx == nullptr) {
+		throw internal_error();
+	}
+	if (!EVP_DecryptInit_ex(ctx, EVP_aes_256_ctr(), nullptr, nullptr, nullptr)) {
+		EVP_CIPHER_CTX_free(ctx);
+		throw internal_error();
+	}
+	if (!EVP_DecryptInit_ex(ctx, nullptr, nullptr, key.getPtr()->data(), iv.data())) {
+		EVP_CIPHER_CTX_free(ctx);
+		throw internal_error();
+	}
+}
+
+StringRef DecryptBlobCipherAes256Ctr::decrypt(unsigned char const* ciphertext,
+                                              const int ciphertextLen,
+                                              const BlobCipherEncryptHeader& header,
+                                              Arena& arena) {
+	TEST(true); // Decrypting data with BlobCipher
+	// validate header flag sanity; for now only one valid header version exists.
+	if (header.flags.headerVersion != EncryptBlobCipherAes265Ctr::ENCRYPT_HEADER_VERSION ||
+	    header.flags.encryptMode != BLOB_CIPHER_ENCRYPT_MODE_AES_256_CTR) {
+		throw internal_error();
+	}
+	// encrypted byte checksum sanity; protection against data bit-rot/flip.
+	BlobCipherChecksum checksum = XXH3_64bits(ciphertext, ciphertextLen);
+	if (checksum != header.checksum) {
+		throw internal_error();
```

Comment:
We may want to have a different error type for when we detect corruption/bit flip, so that higher layers like the storage engine can handle it gracefully/differently? I think internal_error is normally what ASSERT throws.
For example, redwood currently throws a checksum_failed() error when it detects corruption

Reply:
Good idea! I think we can use the reuse `checksum_failed` as the definition claims "A data checksum failed" without specifying the "data" characteristics. Is there a reason we need to extend error definitions with more granular definition such as `encrypt_checksum_failed`?

Reply:
Updated the code to use `checksum_failed` for cases where checksum seems to be corrupted; also introduced `encyrpt_ops_error` error-code to handle cases when OpenSSL AES operations fails.